toolchain/docs/parse.md

```diff
@@ -0,0 +1,917 @@
+      {kind: 'ReturnType', text: '->', subtree_size: 2},
+    {kind: 'FunctionDefinitionStart', text: '{', subtree_size: 7},
+      {kind: 'ReturnStatementStart', text: 'return'},
+      {kind: 'Literal', text: '42'},
+    {kind: 'ReturnStatement', text: ';', subtree_size: 3},
+  {kind: 'FunctionDefinition', text: '}', subtree_size: 11},
+  {kind: 'FileEnd', text: ''},
+]
+```
+
+<!-- prettier-ignore-end -->
+
+In this example, `FileStart`, `FunctionDefinition`, and `FileEnd` are "root"
+nodes for the tree. Function components are children of `FunctionDefinition`.
+
+It's produced in this way because it's an efficient layout to produce with
+vectorized storage, requiring little context to be maintained during parsing.
+Because it's stored in postorder, it's also most efficient to process the parsed
+output in postorder; this affects checking.
+
+The parse tree is printed in postorder by default because it matches how the
+parse tree is expected to be processed within the toolchain , and so can make it
+easier to reason about. However, the `--preorder` flag may be used in contexts
+where a preorder representation would be easier to handle.
+
+## Bracketing inside the tree
+
+The parse tree is designed to be walked in postorder by checking, allowing
+checking to be more efficient. To support this, checking sometimes requires
+context on the meaning of a node when it is encountered.
+
+Each `ParseNodeKind` has either a bracketing node, or a specific child count.
+This helps document and enforce the expected tree structure.
+
+When a bracketing node is indicated, it is the opening bracket: it will always
+be the first child of the parent, and that will be the only time it occurs in
+the parent's children (it may still occur in children of children). When
+checking encounters the opening bracket, this means it can make contextual
+decisions for the later children of the node.
+
+Nodes can also have a specific child count, for example, infix operators always
+have two children: the lhs and rhs expressions. Many nodes have a child count of
+0; this just means they're leaf nodes, and will never have children.
+
+Because the tree structure is always valid, these are treated as contracts. Some
+nodes exist only to be used to construct valid tree structures for invalid
+input, such as `StructFieldUnknown`.
+
+Although each subtree's size is also tracked as part of the node, we're
+currently trying to avoid relying on it and may eliminate it if it turns out to
+be unnecessary and a meaningful cost for the compiler.
+
+## Visual example
+
+To try to explain the transition from code to Parse Tree, consider the
+statement:
+
+```carbon
+var x: i32 = y + 1;
+```
+
+Lexing creates distinct tokens for each syntactic element, which will form the
+basis of the parse tree:
+
+<!--
+These charts are using a hidden "root" node in order to adjust alignment.
+We're trying to keep the "lexed" tokens separate from "parse tree" nodes.
```

Comment:
I find these figures pretty hard to follow, compared to the ones in the original doc. Maybe we should keep them as SVG until someone has time to make Mermaid render them a little more like the originals?

Reply:
Yeah, I understand the maintainability concerns, and themeability is certainly nice, but none of that matters unless the figures are successfully communicating the ideas, and right now I don't know if they are. At least for me, what made the original figures effective was that they worked like frames of animation: items only change position within the frame when the figure is depicting a state change, and the frames are small enough that I can keep several of them on screen and visually scan between them to see the changes. I think it was also crucial that the depiction of the tree made the postorder structure explicit -- for example, on line 361 the main point we make about the following figure is that the `;` is bracketed by `var`, but that isn't visible the new figure unless the `var` is on the far left and the `;` is on the far right. Without those things, I don't know if I would have understood what the figures were trying to tell me.

Maybe we should get a second opinion on this?

Reply:
Sure, adding @zygoloid.

Reply:
BTW, here's some edits that layer the frames more.

Reply:
Note, three different approaches:

- [Original](https://github.com/carbon-language/carbon-lang/blob/a6b3e2ad797e2992610d2c7db55b989981d9ef2d/toolchain/docs/parse.md#visual-example)
- [More overlaid](https://github.com/carbon-language/carbon-lang/blob/4b560db502cf720c25488506297970acbfa9d4a8/toolchain/docs/parse.md#visual-example)
- [With some subtle final changes](https://github.com/carbon-language/carbon-lang/blob/9fb8c55db0d87af3449fb459d15dab209749bac6/toolchain/docs/parse.md#visual-example)

Reply:
Per discussion, ascii art